sympy/matrices/random.py

```diff
@@ -0,0 +1,2525 @@
+        if scalar is None else scalar
+    obj = _eye(dim)
+    start, end = sorted(index)
+    obj[start, start] = scalar
+    for i in range(start + 1, end):
+        obj[i - 1, i] = 1
+        obj[i, i] = scalar
+    return obj
+
+
+def transposition(dim,
+                  index=None,
+                  seed=None):
+    r"""n x n transposition matrix
+
+    Explanation
+    ===========
+
+    A transposition matrix is an identity matrix where two columns (or rows)
+    are swapped. It is a special permutation matrix.
+    Moreover, any permutaion matrix is a product of transposition matrices.
+
+    In two dimensions it is
+
+    .. math::
+
+       \mathbf{T} = \left[\begin{array}{cc}0 & 1 \\ 1 & 0\end{array}\right] \\
+
+    If index is given, it sets which column or row will be swapped.
+
+    Examples
+    ========
+
+    >>> from sympy.matrices.random import rand, transposition
+    >>> rand.seed(1)
+
+    >>> transposition(4)
+    Matrix([
+    [1, 0, 0, 0],
+    [0, 0, 1, 0],
+    [0, 1, 0, 0],
+    [0, 0, 0, 1]])
+
+    >>> transposition(3, (0,1))  # no more random
+    Matrix([
+    [0, 1, 0],
+    [1, 0, 0],
+    [0, 0, 1]])
+
+    Parameters
+    ==========
+    dim : integer
+        dimension n of matrix
+    index : tuple(integer, integer) (optional)
+        coordinates (i,j) of transposition
+    seed : random seed (optional)
+        see :func:`projection`
+
+    See Also
+    ========
+    elementary
+    permutation
+
+    """
+    rng = _rand(seed)
+    index = index or _sample(range(dim), 2, rng=rng)
+    return super_elementary_matrix(dim, index)
+
+
+def permutation(dim,
+                perm=None,
+                seed=None):
+    r"""permutation matrix n x n
+
+    Explamation
+    ===========
+
+    A permutation matrix is a matrix with only 0 or 1 entries
+    and each column and each row consists of only one non-zero entry.
+
+    Such a matrix can be obtained by shuffling the rows
+    of an identiy matrix .i.e ``permutation(dim, perm)`` is eqivalent to
+    ``eye(dim).permute(perm)``.
+
+    Examples
+    ========
+    >>> from sympy.matrices.random import rand, permutation
+    >>> rand.seed(1)
+
+    >>> permutation(3)
+    Matrix([
+    [1, 0, 0],
+    [0, 0, 1],
+    [0, 1, 0]])
+
+    >>> permutation(3, (2,0,1))  # no more random
+    Matrix([
+    [0, 0, 1],
+    [1, 0, 0],
+    [0, 1, 0]])
+
+    Parameters
+    ==========
+    dim : integer
+        dimension n of matrix
+    perm : tuple or list of integer (optional)
+        permutation as list of n integers, e.g. ``[2,1,0,3]`` as a
+        permutation of ``[0,1,2,3]``
+    seed : random seed (optional)
+        see :func:`transposition`
+
+    See Also
+    ========
+    transposition
+
+    """
+    rng = _rand(seed)
+    perm = perm or _sample(range(dim), dim, rng=rng)
+    obj = _eye(dim)
+    for i, j in enumerate(perm):
+        obj[i, i] = 0  # aka zeros(dim)
+        obj[i, j] = 1
+    return obj  # aka _eye(dim).permute(perm)
```

Comment:
Could this just use `eye(dim).permute(perm)`?

Reply:
Could, but would work for `sympy` only. neither `numpy` nor `mpmath` support `eye(dim).permute(perm)`.

Reply:
I have to say I don't really understand the desire to support code here (in the SymPy codebase) that should work with the matrix classes from the other codebases. The obvious way to get a numpy array here would be to compute the sympy Matrix and then convert it to a numpy array at the end. Is there a reason that support for numpy matrices can not just work like that?

The main advantage of using numpy in the internal calculations is lost when most of the code that does the setting of individual elements is done by non-vectorised pure Python code. On the other hand if we were to make performance improvements to sympy's  `permute` functions then this code would not benefit from those because it reimplements them rather than reusing them.

Also how would I even use this function with those other libraries?

Reply:
1. Frist of all, this module wouldn't exists as I started with `numpy`. 
2. Second, using different framework is a good way of testing. If same code works for multiple frameworks, the implementation is kind of generic. 
3. Third, in my opinion the current implementation fits into coding style of the remaining module. The `eye(dim).permute(perm)` approach would be kind of singular compared to the other functions.
4. Forth, a dependency via `eye(dim).permute(perm)` would not be in favour of the module. Assume there will be changes on `eye(dim).permute(perm)`, this doesn't mean `permutation(dim, perm)` should change the same way, too.

To run this code with another framework, just replace all lines before line 120 with your framework wrapper and use `eval`.